### PR TITLE
[GCS] Bookkeeping normal task resources in GCS

### PR DESCRIPTION
**Pull request by Chong-Li**
## Why are these changes needed?
This PR enables GCS to bookkeep each node's normal task resource usage. This will make Ant's GCS-based actor scheduler have a complete view of the cluster resources.

The normal task resources are maintained in the cluster_scheduling_resources_ of GcsResourceManager. The GcsResourceManagerEx, which inherits from GcsResourceManager, manages the updates of normal task resources from Raylets.

## Related issue number

#15943

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by wuisawesome** [comment]
Can't this already be calculated because it's the total - actor - placement group resources?

**Comment by Chong-Li** [comment]
> Can't this already be calculated because it's the total - actor - placement group resources?

Total - actor - PG - normal tasks = idle resources. To get to know each node's idle resources (for better load balancing perhaps), normal task resources are required at GCS side.

**Comment by wuisawesome** [comment]
Sorry, I meant total resource load (which is already exists).

**Comment by Chong-Li** [comment]
> Sorry, I meant total resource load (which is already exists).

If actor scheduling is executed by GCS, then there is probably a gap between an actor scheduling decision (made by GCS) and the decision eventually being enforced at the target node. This could make the total load reported by nodes different from the resource view at the GCS side. So we use the formula (Total resources - actor - PG - normal tasks = idle resources) to calculate idle resources. Actor and PG resources are naturally known by GCS because they are all scheduled by it. Then the only thing GCS needs to know is the normal task resources, which is out of its control because normal tasks are scheduled by Raylets.

File: src/ray/gcs/gcs_server/gcs_resource_manager.h [hunk 1]
```diff
@@ -209,5 +215,29 @@ class GcsResourceManager : public rpc::NodeResourceInfoHandler {
   uint64_t counts_[CountType::CountType_MAX] = {0};
 };
 
+class GcsResourceManagerEx : public GcsResourceManager {
```
Comment (wuisawesome) [inline src/ray/gcs/gcs_server/gcs_resource_manager.h:218]: Can you add a description of what this class is, and why it can't just be part of `GcsResourceManager`?

File: src/ray/raylet/scheduling/cluster_resource_data.h [hunk 2]
```diff
@@ -237,3 +239,7 @@ NodeResources ResourceMapToNodeResources(
 TaskRequest ResourceMapToTaskRequest(
     StringIdMap &string_to_int_map,
     const std::unordered_map<std::string, double> &resource_map);
+
```
Comment (wuisawesome) [inline src/ray/raylet/scheduling/cluster_resource_data.h:242]: Can you move this to `cluster_resource_data` (and ideally make the logic there call this function?)

File: src/ray/raylet/scheduling/cluster_resource_data.h [hunk 3]
```diff
@@ -85,6 +85,8 @@ class TaskRequest {
   bool IsEmpty() const;
   /// Returns human-readable string for this task request.
   std::string DebugString() const;
+
+  TaskRequest &operator+=(const TaskRequest &other);
```
Comment (wuisawesome) [inline src/ray/raylet/scheduling/cluster_resource_data.h:89]: this operator is kinda weird. it doesn't really make sense to add task requests together. can we not do this?

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 4]
```diff
@@ -1096,5 +1096,40 @@ void ClusterTaskManager::SpillWaitingTasks() {
   }
 }
 
+ResourceSet ClusterTaskManager::CalcNormalTaskResources() const {
+  TaskRequest total_normal_task_request;
```
Comment (wuisawesome) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:1100]:
```suggestion
  std::unordered_map<std::string, FixedPoint> total_normal_task_request;
```

File: src/ray/gcs/gcs_server/gcs_resource_manager.h (hunk 1 above)
Comment (Chong-Li) [reply src/ray/gcs/gcs_server/gcs_resource_manager.h:218]: The description will be added as "GcsResourceManagerEx is responsible for updating normal task resources from node heartbeats. It is required by GCS based actor scheduling."

This class is not part of GcsResourceManager because it handles heartbeats and UpdateResourcesRequest in different ways. We could put these logic inside GcsResourceManager but that would require more if.. else.. branches. Meanwhile, we do need a single-purpose heartbeat handling function, which will be passed and called by other logic (e.g., workerLease-failure handler, which will be included in the next PR).

BTW, are you Ok with a class name ending with "Ex"?

File: src/ray/raylet/scheduling/cluster_task_manager.cc (hunk 4 above)
Comment (Chong-Li) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:1100]: Yes, this looks better. In this way, the above "operator+=" and TaskRequestToResourceMap() are no longer needed as well.

File: src/ray/gcs/gcs_server/gcs_resource_manager.h (hunk 1 above)
Comment (Chong-Li) [reply src/ray/gcs/gcs_server/gcs_resource_manager.h]: This logic is about updating normal task resources from node heartbeats. I've moved it to be part of GcsResourceManager.

File: src/ray/common/task/scheduling_resources.cc [hunk 5]
```diff
@@ -733,10 +737,23 @@ void SchedulingResources::DeleteResource(const std::string &resource_name) {
...
 std::string SchedulingResources::DebugString() const {
   std::stringstream result;
+
+  auto resources_available = resources_available_;
+  resources_available.SubtractResources(resources_normal_tasks_);
+
   result << "\n- total: " << resources_total_.ToString();
-  result << "\n- avail: " << resources_available_.ToString();
+  result << "\n- avail: " << resources_available.ToString();
+  result << "\n- normal_task_stolen: " << resources_normal_tasks_.ToString();
```
Comment (raulchen) [inline src/ray/common/task/scheduling_resources.cc:756]: "normal task usage:`? stolen sounds weird.

File: src/ray/common/task/scheduling_resources.h [hunk 6]
```diff
@@ -127,6 +127,12 @@ class ResourceSet {
   /// does not exist.
   FixedPoint GetResource(const std::string &resource_name) const;
 
+  /// Return ture if resource_capacity_ constains the specified resource label.
```
Comment (raulchen) [inline src/ray/common/task/scheduling_resources.h:130]: `Return ture if this resource set constains the specified resource label.`
Do not expose the implementation details in the documentation.

File: src/ray/common/task/scheduling_resources.h [hunk 7]
```diff
@@ -489,13 +495,26 @@ class SchedulingResources {
...
+  void SetNormalTaskResources(const ResourceSet &newset);
+
  private:
   /// Static resource configuration (e.g., static_resources).
   ResourceSet resources_total_;
   /// Dynamic resource capacity (e.g., dynamic_resources).
   ResourceSet resources_available_;
   /// Resource load.
   ResourceSet resources_load_;
+  /// Resources of normal tasks.
```
Comment (raulchen) [inline src/ray/common/task/scheduling_resources.h:516]: `Resources used by normal tasks`

File: src/ray/raylet/node_manager.cc [hunk 8]
```diff
@@ -520,6 +534,7 @@ void NodeManager::FillResourceReport(rpc::ResourcesData &resources_data) {
       gcs_client_->NodeResources().GetLastResourceUsage());
   cluster_resource_scheduler_->FillResourceUsage(resources_data);
   cluster_task_manager_->FillResourceUsage(resources_data);
+  FillNormalTaskResourceUsage(resources_data);
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc:537]: Check `gcs_task_scheduling_enabled`

File: src/ray/gcs/gcs_server/gcs_resource_manager.cc [hunk 9]
```diff
@@ -56,9 +56,29 @@ void GcsResourceManager::HandleUpdateResources(
...
+        GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
+        RAY_LOG(DEBUG) << "Changed resources is empty, return directly, node id = "
+                       << node_id;
+        return;
+      }
+    }
+  } else {
+    for (const auto &entry : request.resources()) {
+      changed_resources->emplace(entry.first, entry.second.resource_capacity());
+    }
```
Comment (raulchen) [inline src/ray/gcs/gcs_server/gcs_resource_manager.cc:81]: This part is not related with this PR?

**Review by raulchen — APPROVED** [review]
LGTM. A few small comments.
